rule/expr.go

```diff
@@ -10,6 +10,8 @@ import (
 // An Expr is a logical expression that can be evaluated to a value.
 type Expr interface {
 	Eval(Params) (*Value, error)
+	PushExpr(e Expr) error
```

Comment:
There is an issue with the public API of this package:
- The Expr interface isn't really representative of the different behaviors we have in this package.
- `Param` and `Value` must implement `PushExpr` and `Finalise` even if they don't need them: they are not operators.
- If a user wants to create the simplest Expr possible:
  - He needs to implement 3 methods
  - Then **his code will panic because he didn't implement the `TypedExpression` interface**
  - He needs to implement 4 methods

I think we need to rework a little bit the public API by identifying:
- what is the minimum behaviour needed for a type to be integrated into this package?
- what are the different components of this package? Should Operator and Operand be part of the public API? If not, what should we provide to the user?
- how can we make sure that none of these solutions makes the code complicated to maintain?

Also, I insist on this: Don't hesitate to change anything that was there before and that is no longer relevant.

Reply:
@asdine yup, I agree.  I already have the feeling that I need to write a "How to add an operator" guide.  Packing `PushExpr` and `Finalise` onto `Expr` I think was a clear mistake, and an obvious first step is to pull `PushExpr` and `Finalise` out into an `Operator` interface.   

After that, most of the "pain" comes from the Go level usage of rules package.  We could make the parser use the Go way of doing things (i.e. use the constructor functions: `Not`, `Or`, etc..) but this will means handling panics again (We don't want those functions to have multiple returned values, otherwise the Go level expression of rules will be hideous!).